samples/RefreshFlow/AngularApp/src/core/auth.service.ts

```diff
@@ -0,0 +1,164 @@
+import { Injectable } from '@angular/core';
+import { Http, Headers, RequestOptions, Response, URLSearchParams } from '@angular/http';
+import { Observable } from 'rxjs/Observable';
+import { Subscription } from 'rxjs/Subscription';
+import { BehaviorSubject } from 'rxjs/BehaviorSubject';
+import { RefreshGrantModel } from './models/refresh-grant-model';
+import { ProfileModel } from './models/profile-model';
+import { AuthStateModel } from './models/auth-state-model';
+import { AuthTokenModel } from './models/auth-tokens-model';
+import { RegisterModel } from './models/register-model';
+import { LoginModel } from './models/login-model';
+
+const jwtDecode = require('jwt-decode');
+
+import 'rxjs/add/operator/map';
+import 'rxjs/add/operator/mergeMap';
+import 'rxjs/add/operator/first';
+import 'rxjs/add/operator/catch';
+import 'rxjs/add/operator/do';
+import 'rxjs/add/operator/filter';
+
+import 'rxjs/add/observable/of';
+import 'rxjs/add/observable/interval';
+import 'rxjs/add/observable/throw';
+
+@Injectable()
+export class AuthService {
+
+    private initalState: AuthStateModel = { profile: null, tokens: null, authReady: false };
+    private authReady$ = new BehaviorSubject<boolean>(false);
+    private state: BehaviorSubject<AuthStateModel>;
+    private refreshSubscription$: Subscription;
+
+    state$: Observable<AuthStateModel>;
+    tokens$: Observable<AuthTokenModel>;
+    profile$: Observable<ProfileModel>;
+    loggedIn$: Observable<boolean>;
+
+    constructor(
+        private http: Http,
+    ) {
+        this.state = new BehaviorSubject<AuthStateModel>(this.initalState);
+        this.state$ = this.state.asObservable();
+
+        this.tokens$ = this.state.filter(state => state.authReady)
+            .map(state => state.tokens);
+
+        this.profile$ = this.state.filter(state => state.authReady)
+            .map(state => state.profile);
+
+        this.loggedIn$ = this.tokens$.map(tokens => !!tokens);
+    }
+
+    init() {
+        return this.startupTokenRefresh()
+            .do(() => this.scheduleRefresh());
+    }
+
+    register(data: RegisterModel): Observable<Response> {
```

Comment:
Yeah that's a bit of a bit argument but to keep it short Angular encourages the use of Observables (so do I) because they're basically a the representation of a "Steam" or "Events/Data/whatever else over time" which is perfect for this and basically everything UI based